app/bundles/LeadBundle/EventListener/PointSubscriber.php

```diff
@@ -46,18 +53,80 @@ public function onTriggerBuild(TriggerBuilderEvent $event): void
+        $stageListItem           = $this->translator->trans('mautic.lead.stage.remove');
+        $choices[$stageListItem] = 0;
+        foreach ($stages as $stage) {
+            $choices[$stage['name']] = $stage['id'];
+        }
+
+        $event->addEvent(
+            'lead.changestage',
+            [
+                'group'           => 'mautic.lead.point.trigger',
+                'label'           => 'mautic.lead.lead.events.changestage',
+                'formType'        => StageType::class,
+                'formTypeOptions' => ['items' => $choices],
+                'eventName'       => PointEvents::TRIGGER_ON_EVENT_EXECUTE,
+            ]
+        );
     }
 
     public function onTriggerExecute(TriggerExecutedEvent $event): void
     {
-        if ('lead.changetags' !== $event->getTriggerEvent()->getType()) {
-            return;
+        if ('lead.changetags' === $event->getTriggerEvent()->getType()) {
+            $this->handelChangeTags($event);
+        } elseif ('lead.changestage' === $event->getTriggerEvent()->getType()) {
+            $this->handelChangeStage($event);
         }
+    }
 
-        $properties = $event->getTriggerEvent()->getProperties();
-        $addTags    = $properties['add_tags'] ?: [];
-        $removeTags = $properties['remove_tags'] ?: [];
+    /**
+     * Add or remove tags from a contact based on the trigger event.
+     */
+    private function handelChangeTags(TriggerExecutedEvent $event): void
+    {
+        $properties         = $event->getTriggerEvent()->getProperties();
+        $addTags            = $properties['add_tags'] ?: [];
+        $removeTags         = $properties['remove_tags'] ?: [];
 
         $this->leadModel->modifyTags($event->getLead(), $addTags, $removeTags);
     }
+
+    /**
+     * Change or remove the stage of a contact based on the trigger event.
+     */
+    private function handelChangeStage(TriggerExecutedEvent $event): void
+    {
+        $stageId = (int) $event->getTriggerEvent()->getProperties()['stage'];
+
+        if (0 === $stageId) {
+            $this->leadModel->removeFromStage(
+                $event->getLead(),
+                null,
+                $this->translator->trans('mautic.stage.event.removed.batch')
+            );
+
+            return;
+        }
+
+        $stage = $this->stageModel->getEntity($stageId);
+        if (null === $stage || !$stage->isPublished()) {
+            $event->setFailed();
+            $this->logger->error("Stage for ID $stageId not found");
+
+            return;
+        }
+
+        try {
+            $this->leadModel->changeStage(
+                $event->getLead(),
+                $stage,
+                $this->translator->trans('mautic.lead.point.trigger')
+            );
+        } catch (\UnexpectedValueException $e) {
+            $this->logger->info("LeadBundle: Stage not updated for lead {$event->getLead()->getId()} by trigger because: {$e->getMessage()}");
```

Comment:
Shouldn't this `setFailed()` as well? And perhaps even `setSuccess()` in the `try` part?

Reply:
added setFailed() and setSuccess()